src/rgw/rgw_aws.cc

```diff
@@ -0,0 +1,303 @@
+    if (outcome.IsSuccess()) {
+      std::cout << "Invoke function: " << request.GetFunctionName() << "with body " << request.GetBody() << std::endl;
+      if(derived && derived->cb) {
+        derived->cb(STATUS_OK);
+      }
+    }
+    else {
+      const auto& error = outcome.GetError();
+      std::cout << "ERROR: " << error.GetExceptionName() << ": "
+                << error.GetMessage() << std::endl;
+      if(derived && derived->cb) {
+        derived->cb(-1);
+      }
+    }
+
+  }
+  void publish_callback(const Aws::SNS::SNSClient* client,
+                        const Aws::SNS::Model::PublishRequest& request,
+                        const Aws::SNS::Model::PublishOutcome & outcome,
+                        const std::shared_ptr<const Aws::Client::AsyncCallerContext>& context){
+
+    auto derived = std::static_pointer_cast<const AsyncCallerContext>(context);
+    if (outcome.IsSuccess()) {
+      std::cout << "SNS: published to topic: " << request.GetTopicArn() << std::endl;
+      if(derived && derived->cb) {
+        derived->cb(STATUS_OK);
+      }
+
+    }
+    else {
+      const auto& error = outcome.GetError();
+      std::cout << "ERROR: " << error.GetExceptionName() << ": "
+                << error.GetMessage() << std::endl;
+      if(derived && derived->cb) {
+        derived->cb(-1);
+      }
+    }
+  }
+
+  struct message_t {
+    const AwsClient client;
+    const std::string resourceARN;
+    const std::string payload;
+    reply_callback_t cb;
+
+    message_t(const AwsClient client, const std::string& resourceARN, const std::string& payload, reply_callback_t cb) :
+            client(client),
+            resourceARN(resourceARN),
+            payload(payload),
+            cb(cb) {};
+  };
+
+  typedef std::unordered_map<client_id_t, AwsClient, client_id_t::hasher> ClientList;
+
+
+  typedef boost::lockfree::queue<message_t *, boost::lockfree::fixed_sized<true>> MessageQueue;
+
+  class Manager {
+  private:
+    MessageQueue messageQueue;
+    ClientList clientList;
+    std::thread runner;
+    bool stopped;
+    mutable std::mutex clientsLock;
+    CephContext *cct;
+    const ceph::coarse_real_clock::duration idle_time;
+    const int MAX_CONNECTIONS = 25;
+    const int REQUEST_TIMEOUT_MS = 3000;
+    const int CONNECT_TIMEOUT_MS = 3000;
+    Aws::SDKOptions options;
+
+    void publish_internal_wrapper(message_t *msg) {
+      const std::unique_ptr<message_t> msg_owner(msg);
+
+      std::visit([&msg, this](auto &&arg) { this->publish_internal(arg, msg->resourceARN, msg->payload, msg->cb); }, msg->client);
+    }
+
+    void publish_internal(Aws::SNS::SNSClient *client, const std::string &resourceArn, const std::string &payload, reply_callback_t cb) {
+      Aws::SNS::Model::PublishRequest req;
+      req.SetMessage(Aws::String(payload));
+      req.SetTopicArn(Aws::String(resourceArn));
+
+      auto context = Aws::MakeShared<AsyncCallerContext>("PublishAllocationTag", cb);
```

Comment:
you should keep a counter (could be an atomic int) which makes sure we don;t have too many pending requests.
e.g. if aws connection is down
if counter exceeds you can threat that as an internal error (don't forget to invoke the cb() with some error code)
this also apply to line 160

Reply:
> you should keep a counter (could be an atomic int) which makes sure we don;t have too many pending requests.
> e.g. if aws connection is down
> if counter exceeds you can threat that as an internal error (don't forget to invoke the cb() with some error code)
> this also apply to line 160

after further discussion with @prateek2211 , no need for that counter. messages will timeout if server is down/slow.